src/librustc_data_structures/tagged_ptr.rs

```diff
@@ -0,0 +1,125 @@
+//! This module implements tagged pointers.
+//!
+//! In order to utilize the pointer packing, you must have two types: a pointer,
+//! and a tag.
+//!
+//! The pointer must implement the `Pointer` trait, with the only requirement
+//! being conversion to and from a usize. The alignment of the Pointee dictates
+//! the space available in the least significant bits of the returned pointers.
+//!
+//! Note that we compare pointers by Pointee rather than the pointers
+//! themselves. This is unfortunate, but we cannot help it.
+//!
+//! Many common pointer types already implement the `Pointer` trait.
+//!
+//! The tag must implement the `Tag` trait; it should fit into the alignment
+//! provided by the `Pointer`.
+
+use std::mem::ManuallyDrop;
+use std::rc::Rc;
+use std::sync::Arc;
+
+mod copy;
+mod drop;
+
+pub use copy::CopyTaggedPtr;
+pub use drop::TaggedPtr;
+
+/// This describes the pointer type encaspulated by TaggedPtr.
+///
+/// # Safety
+///
+/// The usize returned from `into_usize` must be a valid, dereferenceable,
+/// pointer to `Pointee`.
```

Comment:
Presumably there is also a safety contract for `from_usize` and `with_ref` that implementations of the trait need to follow?